test/widgets/message_list_test.dart

```diff
@@ -222,6 +238,87 @@ void main() {
+        final narrow = TopicNarrow(someStream.streamId, eg.t(someTopic), with_: 1);
+        await setupMessageListPage(tester,
+          narrow: narrow,
+          // server sends the /with/<id> message in its current, different location
+          messages: [eg.streamMessage(id: 1, stream: otherStream, topic: otherTopic)],
+          streams: [someStream, otherStream],
+          subscriptions: [eg.subscription(someStream), eg.subscription(otherStream)],
+          skipPumpAndSettle: true);
+        await tester.pump(); // global store loaded
+        await tester.pump(); // per-account store loaded
+
+        // Until we learn the conversation was moved,
+        // we put the link's stream/topic in the app bar.
+        checkAppBarChannelTopic(someStream.name, someTopic);
+
+        await tester.pumpAndSettle(); // initial message fetch plus anything else
+
+        // When we learn the conversation was moved,
+        // we put the new stream/topic in the app bar.
+        checkAppBarChannelTopic(otherStream.name, otherTopic);
+
+        // We followed the move in just one fetch.
+        check(connection.takeRequests()).single.isA<http.Request>()
+          ..method.equals('GET')
+          ..url.path.equals('/api/v1/messages')
+          ..url.queryParameters.deepEquals({
+            'narrow': jsonEncode(narrow.apiEncode()),
+            'anchor': AnchorCode.newest.toJson(),
+            'num_before': kMessageListFetchBatchSize.toString(),
+            'num_after': '0',
+          });
+      });
+
+      testWidgets('without message move', (tester) async {
+        final narrow = TopicNarrow(someStream.streamId, eg.t(someTopic), with_: 1);
+        await setupMessageListPage(tester,
+          narrow: narrow,
+          // server sends the /with/<id> message in its current, different location
+          messages: [eg.streamMessage(id: 1, stream: someStream, topic: someTopic)],
+          streams: [someStream],
+          subscriptions: [eg.subscription(someStream)],
+          skipPumpAndSettle: true);
+        await tester.pump(); // global store loaded
+        await tester.pump(); // per-account store loaded
+
+        // Until we learn if the conversation was moved,
+        // we put the link's stream/topic in the app bar.
+        checkAppBarChannelTopic(someStream.name, someTopic);
+
+        await tester.pumpAndSettle(); // initial message fetch plus anything else
+
+        // There was no move, so we're still showing the same stream/topic.
+        checkAppBarChannelTopic(someStream.name, someTopic);
+
+        // We only made one fetch.
+        check(connection.takeRequests()).single.isA<http.Request>()
+          ..method.equals('GET')
+          ..url.path.equals('/api/v1/messages')
+          ..url.queryParameters.deepEquals({
+            'narrow': jsonEncode([
+              ApiNarrowStream(someStream.streamId),
+              ApiNarrowTopic(eg.t(someTopic)),
+              ApiNarrowWith(1),
+            ]),
```

Comment:
This can use `narrow.apiEncode()` like the other test case does, right? Or is there a difference I'm missing?

Reply:
Oh right! That makes sense.